js/webidl2.js

```diff
@@ -687,6 +687,69 @@
             }
             return ret;
         };
+
+        var iterable_type = function() {
+            if (consume(ID, "iterable")) return "iterable";
+            else if (consume(ID, "legacyiterable")) return "legacyiterable";
+            else if (consume(ID, "maplike")) return "maplike";
+            else if (consume(ID, "setlike")) return "setlike";
+            else return;
+        }
+
+        var readonly_iterable_type = function() {
+            if (consume(ID, "maplike")) return "maplike";
+            else if (consume(ID, "setlike")) return "setlike";
+            else return;
+        }
+
+        var iterable = function (store) {
+            all_ws(store, "pea");
+            var grabbed = [],
+                ret = {type: null, idlType: null, readonly: false};
+            if (consume(ID, "readonly")) {
+                ret.readonly = true;
+                grabbed.push(last_token);
+                var w = all_ws();
+                if (w) grabbed.push(w);
+            }
+            var consumeItType = ret.readonly ? readonly_iterable_type : iterable_type;
+
+            var ittype = consumeItType();
+            if (!ittype) {
+                tokens = grabbed.concat(tokens);
+                return;
+            }
+
+            var secondTypeRequired = ittype === "maplike";
+            var secondTypeAllowed = secondTypeRequired || ittype === "iterable";
+            ret.type = ittype;
+            if (ret.type !== 'maplike' && ret.type !== 'setlike')
+                delete ret.readonly;
+            all_ws();
+            if (consume(OTHER, "<")) {
```

Comment:
Cyclomatic complexity is way too high :( I'm worried about the maintainability of this code. @dontcallmedom, I know you said we should handle this stuff kind of code going forward.